### PR TITLE
Product data structure

### PR DESCRIPTION
**Pull request by MaroderKo**

### CODE REVIEW

File: src/main/java/com/autosale/shop/controller/ProductController.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import com.autosale.shop.service.ProductService;
+import lombok.RequiredArgsConstructor;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+@Controller
+@RequestMapping("/product")
```
Comment (YaroslavMovchan) [inline src/main/java/com/autosale/shop/controller/ProductController.java]: /products

File: src/main/java/com/autosale/shop/model/Product.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+    Integer id;
+    String name;
+    String description;
+    Float cost;
+    ProductStatus status;
+    Integer sellerId;
+    Integer buyerId;
+
+    public Product(@JsonProperty("id") Integer id, @JsonProperty("name") String name, @JsonProperty("description") String description, @JsonProperty("cost") Float cost, @JsonProperty("status") ProductStatus status, @JsonProperty("seller_id") Integer sellerId, @JsonProperty("buyer_id") Integer buyerId) {
```
Comment (YaroslavMovchan) [inline src/main/java/com/autosale/shop/model/Product.java:17]: format your constructor

File: src/main/java/com/autosale/shop/repository/impl/ProductRepositoryImpl.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+    @Override
+    public int deleteById(int id) {
+        return dsl.deleteFrom(PRODUCT)
+                .where(PRODUCT.ID.eq(id))
+                .execute();
+    }
+
+    @Override
+    public List<Product> findAllActive() {
```
Comment (YaroslavMovchan) [inline src/main/java/com/autosale/shop/repository/impl/ProductRepositoryImpl.java]: probably is better to do a universal endpoint as findByStatus?

File: src/main/java/com/autosale/shop/service/impl/ProductServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        return repository.findById(id).orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, String.format("Product with id %d not found", id)));
+    }
+
+    @Override
+    public List<Product> findAllFromCurrentUser() {
+        return repository.findAllByUserId((Integer) getContext().getAuthentication().getPrincipal());
+    }
+
+    @Override
+    public List<Product> findAllFromUser(int id) {
```
Comment (YaroslavMovchan) [inline src/main/java/com/autosale/shop/service/impl/ProductServiceImpl.java]: findAllByUserId

File: src/main/java/com/autosale/shop/service/impl/ProductServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    @Override
+    public Product findById(int id) {
+        return repository.findById(id).orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, String.format("Product with id %d not found", id)));
+    }
+
+    @Override
+    public List<Product> findAllFromCurrentUser() {
+        return repository.findAllByUserId((Integer) getContext().getAuthentication().getPrincipal());
```
Comment (YaroslavMovchan) [inline src/main/java/com/autosale/shop/service/impl/ProductServiceImpl.java]: Can Principal be converted to int?

File: src/main/java/com/autosale/shop/service/impl/ProductServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    public int deleteById(int id) {
+        if (getContext().getAuthentication().getAuthorities().contains(new SimpleGrantedAuthority("ROLE_ADMIN")) || findById(id).getSellerId().equals(getContext().getAuthentication().getPrincipal())) {
+            return repository.deleteById(id);
+        } else {
+            throw new ResponseStatusException(HttpStatus.FORBIDDEN);
+        }
+    }
+
+    @Override
+    public void makeSold(int productId) {
```
Comment (YaroslavMovchan) [inline src/main/java/com/autosale/shop/service/impl/ProductServiceImpl.java]: buy?

Should it be synchronized in some way? Why if some users buy the same product in the same time?

File: src/main/java/com/autosale/shop/service/impl/ProductServiceImpl.java (hunk 5 above)
Comment (MaroderKo) [reply src/main/java/com/autosale/shop/service/impl/ProductServiceImpl.java]: If it's a question about "Can I simplify Integer to int?" then yes and I'll do it, and if it's a question about "Is principal is int inside?" then again yes, when I create authentication in security filter I use user id, which is integer, as authentication principal

File: src/main/java/com/autosale/shop/service/impl/ProductServiceImpl.java (hunk 6 above)
Comment (MaroderKo) [reply src/main/java/com/autosale/shop/service/impl/ProductServiceImpl.java]: added modificator synchronized to method

File: src/main/java/com/autosale/shop/repository/impl/ProductRepositoryImpl.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+                .where(PRODUCT.ID.eq(id))
+                .execute();
+    }
+
+    @Override
+    public List<Product> findAllWithConditions(Pagination pagination, List<Condition> conditions) {
+        return dsl.selectFrom(PRODUCT)
+                .where(conditions)
+                .offset((pagination.getCurrentPage() - 1) * pagination.getPageSize()) // -1 because we start pages from 1
+                .fetchSize(pagination.getPageSize())
```
Comment (YaroslavMovchan) [inline src/main/java/com/autosale/shop/repository/impl/ProductRepositoryImpl.java]: use limit

File: src/main/java/com/autosale/shop/controller/ProductController.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+public class ProductController {
+    private final ProductService service;
+
+    @GetMapping()
+    public ResponseEntity<List<Product>> getAll(@RequestBody Pagination pagination) {
+        return ResponseEntity.ok(service.findAll(pagination));
+    }
+
+    @GetMapping("/status/{status}")
+    public ResponseEntity<List<Product>> getAllByStatus(@RequestBody Pagination pagination, @PathVariable String status) {
```
Comment (YaroslavMovchan) [inline src/main/java/com/autosale/shop/controller/ProductController.java]: 1. pass status as Enum here
2. /statuses/{status}

**Pull request merged**